core/src/main/java/io/grpc/internal/MessageDeframer.java

```diff
@@ -438,6 +449,18 @@ public synchronized void reset() throws IOException {
       count = mark;
     }
 
+    @Override
+    public void close() throws IOException {
+      super.close();
+    }
+
+    private void reportCount() {
+      if (count > maxCount) {
+        statsTraceContext.messageReceived(count - maxCount);
```

Comment:
This is problematic. I wanted to have `messageReceived()` called once for each message, so that if tracing will want to expose message reading events, it will get it right. Now it may be recorded as multiple messageReceived events, depending on how fine-grained it is read.

Alternatively, I could do this in `close()`, but that requires all Marshallers to close the InputStream. No Marshaller does it today. This would be an API change. Not that it would break users, but that stats won't work correctly if the implementation doesn't follow the new API. I think it should be fine. @ejona86 WDYT?